cpu/qn908x/periph/rtc.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (C) 2020 iosabi
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_qn908x
+ * @ingroup     drivers_periph_rtc
+ *
+ * @{
+ *
+ * @file
+ * @brief       Low-level Real-Time Clock (RTC) driver implementation
+ *
+ * @author      iosabi <iosabi@protonmail.com>
+ *
+ * @}
+ */
+
+
+#include <stdlib.h>
+
+#include "cpu.h"
+#include "board.h"
+#include "periph_conf.h"
+#include "periph/rtc.h"
+
+#include "vendor/drivers/fsl_clock.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+
+/* Callback context. */
+typedef struct {
+    rtc_alarm_cb_t cb;
+    void *arg;
+    uint32_t alarm;
+} rtc_ctx_t;
+
+static rtc_ctx_t rtc_ctx = { NULL, NULL, 0 };
+
+void rtc_init(void)
+{
+    DEBUG("rtc_init(), rtc=%" PRIu32 "\n", RTC->SEC);
+    /* The RTC is really meant to be using the RCO32K at 32KHz. However, if the
+     * 32K clock source is set to the external XTAL at 32.768 KHz we can use a
+     * calibration parameter to correct for this:
+     *  ppm = (32000 / 32768 - 1) * (1 << 20) = -24576
+     */
+    const uint32_t freq_32k = CLOCK_GetFreq(kCLOCK_32KClk);
+
+    if (freq_32k == 32000) {
```

Comment:
You might save a few bytes by determining this at compile-time.

Reply:
Yes, we could save some bytes on init, not sure how important is that in this chip. It is technically possible to disable the external 32K crystal and continue running the internal RCO32 only for a low power power-down-0 mode (the mode with 32K clock running); but for the 32K is maybe useless, I measured only a few uA at 1.8V in that mode, plus I don't think the internal oscillators work in this chip at 1.8v. I'll set this to static values.

For other drivers I compute the clocks at runtime because it is more reasonable to slow down your APB or AHB bus to consume less even with the CPU up. Things are gonna get more exciting once I figure out the `pm` module here (spoiler alert: it won't be pretty).

I was thinking to propose a new API in rtc.h to adjust the clock by doing clock slew. This CPU supports it directly here and that's a better way to sync time. Maybe I should file a FR for that.